LanguageFeatures/Patterns/type_inference_A18_t04.dart

```diff
@@ -0,0 +1,39 @@
+// Copyright (c) 2023, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+/// @assertion Calculate the static type of the pattern. Using that value type,
+/// recurse through the pattern again downwards to the leaf subpatterns filling
+/// in any missing types in the pattern. This process may also insert implicit
+/// coercions and casts from dynamic when values flow into a pattern during
+/// matching.
+///
+/// @description Check that the calculation of the static type of a map pattern
+/// performs casts from dynamic and generic function instantiation
+/// @author sgrekhov22@gmail.com
+
+// SharedOptions=--enable-experiment=patterns
+
+import "../../Utils/expect.dart";
+import "../../Utils/static_type_helper.dart";
+
+T foo<T>(T t) => t;
+
+main() {
+  dynamic pi = 3.14;
+  final <String, double>{"key1": x1} = <String, double>{"key1": pi};
+  x1.expectStaticType<Exactly<double>>();
+  Expect.equals(3.14, x1);
+
+  var {"key1": double x2} = <String, double>{"key1": pi};
+  x2.expectStaticType<Exactly<double>>();
+  Expect.equals(3.14, x1);
+
+  Expect.throws(() {
+    var <String, int>{"key1": x3} = <String, int>{"key1": pi};
+  });
+
+  Map<String, int Function(int)> map = {"key1": foo};
+  final <String, int Function(int)>{"key1": x4} = map;
```

Comment:
This case is precisely the topic of dart-lang/language#2845.

How can we keep this test in sync with the language? Presumably, the language team will discuss 2845 for a while and then conclude something, and at that point we will know what to expect here.

I tend to prefer that we _don't_ support "late coercions" (that is, coercions applied to implicit getter invocations that arise by desugaring of pattern declarations and pattern assignments), but we'll see. It's convenient (probably, at least), but I'm afraid it isn't very good for code comprehensibility and readability.

Reply:
Would it work for you to put a "star" on that issue 2845? I would expect that you will then get notification as the discussions proceed and then it should be possible to return to this particular test when it's known what it should do.

It might also make sense to put this test into a different library, because the topic, "late coercions", could justify a handful of different test cases.

(Coercions: generic function instantiation and dynamic type check on expression of type `dynamic`; cases: coercion occurs at top level, coercion occurs for variable pattern as field in record, coercion occurs on variable pattern as element in list, as value in map; coercion occurs on a subpattern which is not a variable pattern; and stuff like that).

Reply:
I subscribed to 2845 and made a note to myself "update LanguageFeatures/Patterns/type_inference_A18_t04.dart depending on https://github.com/dart-lang/language/issues/2845"

Reply:
Sounds good, thanks!